lld/ELF/LinkerScript.cpp

```diff
@@ -1518,3 +1513,50 @@ void LinkerScript::checkFinalScriptConditions() const {
       checkMemoryRegion(lmaRegion, sec, sec->getLMA());
   }
 }
+
+// Add symbols referred by the provide symbol to the symbol table.
+// This function must only be called for provide symbols that should be added
+// to the link.
+static void
+addProvideSymReferences(StringRef provideSym,
+                        llvm::StringSet<> &addedRefsFromProvideSym) {
+  assert(LinkerScript::shouldAddProvideSym(provideSym) &&
```

Comment:
You can define the static function as a lambda in `addScriptReferencedSymbolsToSymTable` instead so that you can just reference the local variable `addedRefsFromProvideSym`.

Then this assert doesn't seem that useful since the code is very short.

Reply:
This static function is recursive. Lambdas cannot be recursive without using more complicated mechanism such as `std::function` or adding an explicit parameter that would refer to the lambda in the lambda call operator. These techniques are explained here: [Recursive lambda functions in C++14](https://stackoverflow.com/questions/18085331/recursive-lambda-functions-in-c14)

Please give your suggestion regarding what should we do here. `std::function` has performance cost due to the use of type-erasure and passing the lambda to the lambda call (for ex, `L(L, x, ...)`) leads to (slightly) ugly code.

Reply:
I am writing a version that uses a stack.

Reply:
See https://github.com/MaskRay/llvm-project/tree/my-provide
Feel free to cherry pick it into this patch

Reply:
Thank you for providing the stack-based implementation. I have cherry-picked your commit into this patch.